src/lib/jsonParseV2/parseEntitiesVector.cpp

```diff
@@ -0,0 +1,217 @@
+        }
+        if (forbiddenIdCharsV2(idOpt.value.c_str()))
+        {
+          *errorStringP = "forbidden characters in /entities/ vector item /id/";
+          return false;
+        }
+        if (idOpt.value.length() > MAX_ID_LEN)
+        {
+          *errorStringP = "max id length exceeded in /entities/ vector item /id/";
+          return false;
+        }
+        id = idOpt.value;
+      }
+    }
+
+    {
+      Opt<std::string> idPatOpt = getStringOpt(*iter, "idPattern", "/entities/ vector item /idPattern/");
+
+      if (!idPatOpt.ok())
+      {
+        *errorStringP = idPatOpt.error;
+        return false;
+      }
+      else if (idPatOpt.given)
+      {
+        if (idPatOpt.value.empty())
+        {
+          *errorStringP = "/entities/ vector item /idPattern/ is empty";
+          return false;
+        }
+
+        idPattern = idPatOpt.value;
+
+        // FIXME P5: Keep the regex and propagate to sub-cache
+        regex_t re;
+        if (regcomp(&re, idPattern.c_str(), REG_EXTENDED) != 0)
+        {
+          *errorStringP = ERROR_DESC_BAD_REQUEST_INVALID_REGEX_ENTIDPATTERN;
+          return false;
+        }
+        regfree(&re);  // If regcomp fails it frees up itself
+      }
+    }
+
+    {
+      Opt<std::string> typeOpt = getStringOpt(*iter, "type", "/entities/ vector item type");
+
+      if (!typeOpt.ok())
+      {
+        *errorStringP = typeOpt.error;
+        return false;
+      }
+      else if (typeOpt.given)
+      {
+        if (forbiddenIdCharsV2(typeOpt.value.c_str()))
+        {
+          *errorStringP = "forbidden characters in /entities/ vector item /type/";
+          return false;
+        }
+        if (typeOpt.value.length() > MAX_ID_LEN)
+        {
+          *errorStringP = "max type length exceeded in /entities/ vector item /type/";
+          return false;
+        }
+        if (typeOpt.value.empty())
+        {
+          *errorStringP = ERROR_DESC_BAD_REQUEST_EMPTY_ENTTYPE;
+          return false;
+        }
+        type = typeOpt.value;
+      }
+    }
+
+    {
+      Opt<std::string> typePatOpt = getStringOpt(*iter, "typePattern", "/entities/ vector item /typePattern/");
+
+      if (!typePatOpt.ok())
+      {
+        *errorStringP = typePatOpt.error;
+        return false;
+      }
+      else if (typePatOpt.given)
+      {
+        if (typePatOpt.value.empty())
+        {
+          *errorStringP = "/entities/ vector item /typePattern/ is empty";
+          return false;
+        }
+
+        typePattern = typePatOpt.value;
+
+        // FIXME P5: Keep the regex and propagate to sub-cache
+        regex_t re;
+        if (regcomp(&re, typePattern.c_str(), REG_EXTENDED) != 0)
+        {
+          *errorStringP = ERROR_DESC_BAD_REQUEST_INVALID_REGEX_ENTTYPEPATTERN;
+          return false;
+        }
+        regfree(&re);  // If regcomp fails it frees up itself
+      }
+    }
+
+    ngsiv2::EntID  eid(id, idPattern, type, typePattern);
+
+    eivP->push_back(eid);
```

Comment:
I understand this was the point in which the checking we were discussiong was placed. In fact, previously line 212 looked like:

```
    if (std::find(eivP->begin(), eivP->end(), eid) == eivP->end())  // if not already included
    {
      eivP->push_back(eid);
    }
```

Now I have the context, let me think a bit about it...

Reply:
It seems that this checking is to avoid inserting the same entity twice, e.g. if you have something like in the payload:

```
[ {"id": "E1", "type": "T1"}, {"id": "E2", "type": "T2"}, {"id": "E1", "type": "T1"} ]
```

You get a [E1/T1, E2/T2] vector instead of having duplicated E1/T1.

In my opinion, the check is still valid. We should try look a solution to the compillation problem you have with it. Could you paste the compiler message you get in this comment thread to have a look?

Reply:
```
[ 51%] Building CXX object src/lib/jsonParseV2/CMakeFiles/jsonParseV2.dir/parseEntitiesVector.cpp.o
make[3]: Entering directory '/home/kz/git/apinf-orion/BUILD_DEBUG'
Scanning dependencies of target rest
In file included from /usr/include/c++/6/bits/stl_algobase.h:71:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/bits/predefined_ops.h: In instantiation of 'bool __gnu_cxx::__ops::_Iter_equals_val<_Value>::operator()(_Iterator) [with _Iterator = __gnu_cxx::__normal_iterator<ngsiv2::EntID*, std::vector<ngsiv2::EntID> >; _Value = const ngsiv2::EntID]':
/usr/include/c++/6/bits/stl_algo.h:120:14:   required from '_RandomAccessIterator std::__find_if(_RandomAccessIterator, _RandomAccessIterator, _Predicate, std::random_access_iterator_tag) [with _RandomAccessIterator = __gnu_cxx::__normal_iterator<ngsiv2::EntID*, std::vector<ngsiv2::EntID> >; _Predicate = __gnu_cxx::__ops::_Iter_equals_val<const ngsiv2::EntID>]'
/usr/include/c++/6/bits/stl_algo.h:161:23:   required from '_Iterator std::__find_if(_Iterator, _Iterator, _Predicate) [with _Iterator = __gnu_cxx::__normal_iterator<ngsiv2::EntID*, std::vector<ngsiv2::EntID> >; _Predicate = __gnu_cxx::__ops::_Iter_equals_val<const ngsiv2::EntID>]'
/usr/include/c++/6/bits/stl_algo.h:3792:28:   required from '_IIter std::find(_IIter, _IIter, const _Tp&) [with _IIter = __gnu_cxx::__normal_iterator<ngsiv2::EntID*, std::vector<ngsiv2::EntID> >; _Tp = ngsiv2::EntID]'
/home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:213:50:   required from here
/usr/include/c++/6/bits/predefined_ops.h:199:17: error: no match for 'operator==' (operand types are 'ngsiv2::EntID' and 'const ngsiv2::EntID')
  { return *__it == _M_value; }
           ~~~~~~^~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:67:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/bits/stl_iterator.h:856:5: note: candidate: template<class _IteratorL, class _IteratorR, class _Container> bool __gnu_cxx::operator==(const __gnu_cxx::__normal_iterator<_IteratorL, _Container>&, const __gnu_cxx::__normal_iterator<_IteratorR, _Container>&)
     operator==(const __normal_iterator<_IteratorL, _Container>& __lhs,
     ^~~~~~~~
/usr/include/c++/6/bits/stl_iterator.h:856:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/6/bits/stl_algobase.h:71:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/bits/predefined_ops.h:199:17: note:   'ngsiv2::EntID' is not derived from 'const __gnu_cxx::__normal_iterator<_IteratorL, _Container>'
  { return *__it == _M_value; }
           ~~~~~~^~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:67:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/bits/stl_iterator.h:863:5: note: candidate: template<class _Iterator, class _Container> bool __gnu_cxx::operator==(const __gnu_cxx::__normal_iterator<_Iterator, _Container>&, const __gnu_cxx::__normal_iterator<_Iterator, _Container>&)
     operator==(const __normal_iterator<_Iterator, _Container>& __lhs,
     ^~~~~~~~
/usr/include/c++/6/bits/stl_iterator.h:863:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/6/bits/stl_algobase.h:71:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/bits/predefined_ops.h:199:17: note:   'ngsiv2::EntID' is not derived from 'const __gnu_cxx::__normal_iterator<_Iterator, _Container>'
  { return *__it == _M_value; }
           ~~~~~~^~~~~~~~~~~
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/ext/new_allocator.h:139:5: note: candidate: template<class _Tp> bool __gnu_cxx::operator==(const __gnu_cxx::new_allocator<_Tp>&, const __gnu_cxx::new_allocator<_Tp>&)
     operator==(const new_allocator<_Tp>&, const new_allocator<_Tp>&)
     ^~~~~~~~
/usr/include/c++/6/ext/new_allocator.h:139:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/c++/6/bits/stl_algobase.h:71:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/kz/git/apinf-orion/src/lib/jsonParseV2/parseEntitiesVector.cpp:27:
/usr/include/c++/6/bits/predefined_ops.h:199:17: note:   'ngsiv2::EntID' is not derived from 'const __gnu_cxx::new_allocator<_Tp>'
  { return *__it == _M_value; }
           ~~~~~~^~~~~~~~~~~
make[3]: Leaving directory '/home/kz/git/apinf-orion/BUILD_DEBUG'
make[3]: Entering directory '/home/kz/git/apinf-orion/BUILD_DEBUG'
[ 51%] Building CXX object src/lib/rest/CMakeFiles/rest.dir/rest.cpp.o
src/lib/jsonParseV2/CMakeFiles/jsonParseV2.dir/build.make:326: recipe for target 'src/lib/jsonParseV2/CMakeFiles/jsonParseV2.dir/parseEntitiesVector.cpp.o' failed
```

Reply:
I'll have a look

Reply:
Question: this never worked? Or only when the check has been moved to other place?

In other words, if you compile telefonicaid:master (which have the check in the old place) you are able to compile?

Reply:
Well ... of course it worked before. The function changed, slightly, and I got these "hard to understand" errors ... That's when I started to look at what that line really did. I did not find it very interesting and pretty cryptic.

In my opinion, if a client makes a stupid question, it's ok to give a stupid answer ...

We can't possibly cover all possible stupid requests, the broker is very generic in this sense, so why bother about one single case ... ?

That was my reasoning behind removing this check. Not that I think the check itself is stupid.

Reply:
Not all the stupid things the user may do involved the same checking complexity. In this case, the check is easy so I don't see any problem in keeping it. Moreover, removing it implies an slight change in behaviour that is better not to take (somehow related with backward compability, alghouth it is not a mayor case).

All that say if we at end are able to solve the compilation error, of course. I'll have a look to it and keep informed.

Reply:
I think I have solved. It's a piece of cake if you pay attention to the `error: no match for 'operator=='` traced, buried along all the other stuff :)

It is a matter of moving the pair of `operator==()` and `operator!=()` functions for EntID object from Subscription.h to EntID.h file. I'll send you the detail by email.

Reply:
Fixed in 1a49367